Create a derivative for the creation of path alias for entities

nlisgo opened the pull request:
The d.o issue is: https://www.drupal.org/node/2471447

For the moment I have not changed \Drupal\Tests\rules\Integration\Action\NodePathAliasCreateTest much. I will continue to extend the tests for this so it is not dependant on the node entity type.

I invite feedback now on the code for \Drupal\rules\Plugin\Action\EntityPathAliasCreateDeriver and \Drupal\rules\Plugin\Action\EntityPathAliasCreate

nlisgo committed:
create_a_derivative_for-2471447-3.patch

fago commented on src/Plugin/Action/EntityPathAliasCreateDeriver.php:
```diff
@@ -0,0 +1,79 @@
+        'category' => $entity_type->getLabel(),
+        'entity_type_id' => $entity_type_id,
+        'context' => [
+          'entity' => new ContextDefinition("entity:$entity_type_id", $entity_type->getLabel(), $this->t('The entity for which to create a path alias.')),
```
Let's replace "entity" here as well.

nlisgo replied:
This has been addressed now.

nlisgo committed:
Add testing for entity path alias create

nlisgo committed:
Some clean up

nlisgo committed:
Address feedback from @calin-marian

nlisgo commented:
All of the comments from @calin-marian have now been addressed.

Please review.

fubhy commented on tests/src/Integration/Action/EntityPathAliasCreateTest.php:
```diff
@@ -0,0 +1,192 @@
+    ]);
+
+    // Prepare mocked entity manager.
+    $this->entityManager = $this->getMockBuilder('Drupal\Core\Entity\EntityManager')
```
Why are you mocking this yourself? The RulesEntityIntegrationTestBase already does that for you.

fubhy commented on tests/src/Integration/Action/EntityPathAliasCreateTest.php:
```diff
@@ -0,0 +1,192 @@
+    $this->entityManager
+      ->expects($this->any())
+      ->method('getDefinitions')
+      ->will($this->returnValue(['test' => $this->entityType]));
```
Everything you are mocking here for entity manager already exists in parent::setup().

fubhy commented on tests/src/Integration/Action/EntityPathAliasCreateTest.php:
```diff
@@ -0,0 +1,192 @@
+      ->with($this->anything())
+      ->will($this->returnValue(['test' => ['label' => 'Test']]));
+
+    $this->container->set('entity.manager', $this->entityManager);
```
Everything you are mocking here for the EntityManager already exists in parent::setup().

fubhy commented:
Other than that this looks to be ready to go. Can you fix the setup() method (removing the entityManager mocking and the mocking of the associated methods on the entity manager?

nlisgo commented:
@fubhy thanks for the review. I will work on this today.

nlisgo committed:
Addressing feedback from @fubhy

nlisgo commented:
@fubhy I've addressed your feedback and more. I was also able to determine that the entity_test module is enabled in RulesEntityIntegrationTestBase so we have an entity type (entity_test) that we can use removing the need to create a new one.

fubhy commented on src/Plugin/Action/EntityPathAliasCreateDeriver.php:
```diff
@@ -0,0 +1,79 @@
+        continue;
+      }
+
+      $this->derivatives["entity:$entity_type_id"] = [
```
Not every entity type supports the path field. You need to check if the current entity type in this loop supports it and only then add it.

klausi commented on src/Plugin/Action/EntityPathAliasCreate.php:
```diff
@@ -43,7 +32,14 @@ class NodePathAliasCreate extends RulesActionBase implements ContainerFactoryPlu
+  protected $entityTypeId;
+
+  /**
+   * Constructs an EntityCreate object.
```
Wrong class name here.

klausi commented on src/Plugin/Action/EntityPathAliasCreateDeriver.php:
```diff
@@ -0,0 +1,79 @@
+  public function getDerivativeDefinitions($base_plugin_definition) {
+    foreach ($this->entityManager->getDefinitions() as $entity_type_id => $entity_type) {
+      // Only allow content entities with a path field.
+      if (!$entity_type instanceof ContentEntityTypeInterface || !isset($this->entityManager->getBaseFieldDefinitions($entity_type_id)['path'])) {
```
Why do we restrict this to entities with a path field? According to EntityInterface every entity has a URL, so we can also create a path alias for any entity. I think we can simply remove the second part of the condition.

klausi commented on tests/src/Integration/Action/EntityPathAliasCreateTest.php:
```diff
@@ -0,0 +1,217 @@
+   *
+   * @var \PHPUnit_Framework_MockObject_MockObject|\Drupal\Core\Field\BaseFieldDefinition
+   */
+  protected $baseFieldDefinition;
```
I guess this should not be necessary anymore once you don't access the base fields in the deriver anymore.

calin-marian commented:
fubhy said in it's previous comment that not all entity types support the path field. Therefore he suggested to check for the field.
see: https://github.com/fago/rules/pull/164#discussion-diff-28599194

klausi commented:
Since the path field is only a widget masking the alias storage we don't have to care at all about it :-)

nlisgo committed:
Reference the correct class name

calin-marian commented:
In this case reverting the previous commit would get us to the desired state. With the exception of some "node" keywords that were skipped.

If nlisgo can't find the time to work on this, I can take over in the weekend.

nlisgo commented:
@calin-marian I have taken that approach and reverted to the previous commit. Please tell me more about the "node" keywords that were skipped.

calin-marian commented on tests/src/Integration/Action/EntityPathAliasCreateTest.php:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  /**
+   * Tests the action execution with an unsaved node.
```
I thought there were 2 instances, but it's only one. Instead of "Tests the action execution with an unsaved node" it should be "Tests the action execution with an unsaved entity".
This is in tests/src/Integration/Action/EntityPathAliasCreateTest.php at line 59

nlisgo committed:
Correct documentation

klausi commented:
Merged, thanks!

calin-marian commented:
Yay, congratulations Nathan !!!

nlisgo commented:
Great. I'll pick up another issue soon.
